File: collatz-conjecture.py
# This replicates the Collatz conjucture where you take any number,
#  if it's even you divide by 2, if it's odd you multiply by 3 and add 1, 
#  and you repeat the process till you get 0

import os
import matplotlib.pyplot as plt

random_Number = int(input("Enter the initial value: "))
values = []
loop_count = 0

while random_Number != 1:
#  1 is used above to create an output, the correct number is 0 which is an infinite loop
    loop_count += 1
    if random_Number % 2 == 0:
        even = random_Number // 2
        values.append(even)
        random_Number = even 
    else:
        odd = random_Number * 3 + 1
        values.append(odd)
        random_Number = odd 

valuesString = '  ,  '.join(map(str, values))

# Write the values to a text file
with open('values.txt', 'w') as file:
    file.write(valuesString)

print('Values have been written to values.txt')

# Sort the array in descending order
sorted_values = sorted(values, reverse=True)

# Get the five largest numbers
five_largest = sorted_values[:5]

print("The five largest numbers are:", five_largest)

# Create an array of positions for the x-axis
positions = list(range(1, len(values) + 1))

# Plot the chart
plt.plot(positions, values, marker='o')
plt.xlabel('Position in Loop')
plt.ylabel('Number')
plt.title('Chart of Values')
plt.grid(True)
plt.show()

print("The loop ran",loop_count,"times")

File: collatz-conjecture.js
// Collatz Conjecture

const fs = require('fs');

let random_Number = 300005000;
let values = [];

while (random_Number !== 1) { 
// 1 is used above to create an output, the correct number is 0 which is an infinite loop
        if (random_Number % 2 === 0) {
            let even = random_Number / 2;
            random_Number = even;
            values.push(random_Number);
        } else {
            let odd = random_Number * 3 + 1;
            random_Number = odd;
            values.push(random_Number);
        }
}
// console.log(values);

// Convert the array to a string representation
const valuesString = values.join('  ,  ');

// Clear the contents of the "values.txt" file before writing new data
fs.truncate('values-js.txt', 0, (err) => {
    if (err) {
        console.error('Error clearing the file:', err);
    } else {
        // Write the values to a text file
        fs.writeFile('values-js.txt', valuesString, (err) => {
            if (err) {
                console.error('Error writing to the file:', err);
            } else {
                console.log('Values have been written to values-js.txt');
            }
        });
    }
});

// Sort the array in descending order
let sortedValues = values.slice().sort((a, b) => b - a);

// Get the five largest numbers
let fiveLargest = sortedValues.slice(0, 5);

console.log("The five largest numbers are:", fiveLargest);